[JIT x RPC] Consolidate Future type class and Future impl class

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40406 [JIT x RPC] Consolidate Future type class and Future impl class**

Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Reference, https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

xush6528 committed:
[JIT x RPC] Consolidate Future type class and Future impl class

Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

[ghstack-poisoned]

xush6528 committed:
Update on "[JIT x RPC] Consolidate Future type class and Future impl class"


Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

[ghstack-poisoned]

mrshenli commented on torch/_jit_internal.py:
```diff
@@ -650,26 +651,25 @@ def safe_is_subclass(the_type, super_type):
+        raise RuntimeError(
+            "Attempted to use torch.jit.Future without a "
+            "contained type. Please add a contained type, e.g. "
+            "torch.jit.Future[int]"
```
Isn't the `Future` type above deleted? Should this be `torch.futures.Future`?

wanchaol replied:
I think because `torch.jit` still imports Future from this file so it still alive. we shouldn't delete it at least I think, because of BC.

xush6528 replied:
`torch.jit.Future` and `torch.futures.Future` both work as type class after this PR.

Let me change it to a bare `Future` to avoid showing bias to users.

mrshenli commented on torch/_jit_internal.py:
```diff
@@ -650,26 +651,25 @@ def safe_is_subclass(the_type, super_type):
+            raise RuntimeError(
+                "Attempted to use torch.jit.RRef without a "
+                "contained type. Please add a contained type, e.g. "
+                "torch.jit.RRef[int]"
```
ditto

wanchaol replied:
For this it's fine to use `torch.distributed.rpc.RRef` as no BC issue involves.

xush6528 replied:
Let me change it to a bare `RRRef` to avoid showing bias to users.

mrshenli commented on docs/source/futures.rst:
```diff
@@ -18,7 +18,7 @@ on :class:`~torch.futures.Future` objects. Currently, the
 
 .. autoclass:: Future
-    :members:
+    :inherited-members:
```
Let's make sure it does not show `PyFuture` in the doc

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -138,3 +143,21 @@ def wait_all(futures):
+    class FutureMeta(PyFuture.__class__, GenericWithOneTypeVar.__class__):
+        pass
+
+    class Future(PyFuture, GenericWithOneTypeVar, metaclass=FutureMeta):
```
Curious, why not always use this approach instead of first trying `class Future(PyFuture, GenericWithOneTypeVar)`? Is there any downside of the fallback path?

xush6528 replied:
These 2 paths are actually exclusive to each other. I haven't figured out what is the good test condition to decide which path to use.

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -1,59 +1,63 @@
     rref.to_here(5.0).append(6)
 
-def return_value(value):
-    # type: (int) -> int
```
Is this deprecated? If so, we will also need to modify docs

xush6528 replied:
Yes, python2 type comment should be deprecated, since we only support Python3 now.

mrshenli reviewed (approved):
LGTM! Left some minor comments inline. Shall we land after branch cut today?

wanchaol commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -785,7 +772,7 @@ def nonexisting_script():
         @torch.jit.script
         def rpc_async_call_remote_nonexisting_torchscript_in_torchscript(
-            dst_worker_name: str
+            dst_worker_name: str,
```
nit: as they are just single arg, what's those comma for?

xush6528 replied:
These are auto-formatter generated changes. It always append comma for the last argument.

wanchaol reviewed (approved):
looks good.

xush6528 committed:
Update on "[JIT x RPC] Consolidate Future type class and Future impl class"


Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Reference, https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

[ghstack-poisoned]

xush6528 committed:
Update on "[JIT x RPC] Consolidate Future type class and Future impl class"


Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Reference, https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

[ghstack-poisoned]

xush6528 committed:
Update on "[JIT x RPC] Consolidate Future type class and Future impl class"


Same motivation for https://github.com/pytorch/pytorch/issues/35110.

`Future` and `RRef` are two important types for `rpc` module, should make users feel easy to use.

Follow https://github.com/pytorch/pytorch/pull/35694.

Reference, https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

Differential Revision: [D7722176](https://our.internmc.facebook.com/intern/diff/D7722176/)

[ghstack-poisoned]
